libgnucash/app-utils/gnc-optiondb.cpp

```diff
@@ -0,0 +1,1281 @@
+{
+    auto str{kvp->get<const char*>()};
+    if (option.get_ui_type() == GncOptionUIType::BOOLEAN)
+        option.set_value(*str == 't' ? true : false);
+    else
+        option.set_value(std::string{str});
+}
+
+static inline void
+fill_option_from_guid_kvp(GncOption& option, KvpValue* kvp)
+{
+    auto guid{kvp->get<GncGUID*>()};
+    option.set_value(
+        (const QofInstance*)qof_instance_from_guid(guid, option.get_ui_type()));
+}
+
+void
+GncOptionDB::load_from_kvp(QofBook* book) noexcept
+{
+    foreach_section(
+        [book](GncOptionSectionPtr& section)
+        {
+            section->foreach_option(
+                [book, &section](GncOption& option)
+                {
+                    // Make path list as above.
+                    std::string name;
+                    /* qof_book_set_option wants a GSList path. Let's avoid
+                     * allocating and make one here. */
+                    GSList list_tail{}, list_head{nullptr, &list_tail};
+                    if (strcmp(section->get_name().c_str(), "Counters") == 0)
+                        counter_option_path(option, &list_head, name);
+                    else
+                        option_path(option, &list_head);
+                    auto kvp = qof_book_get_option(book, &list_head);
+                    if (!kvp)
+                        return;
+                    switch (kvp->get_type())
+                    {
+                        case KvpValue::Type::DOUBLE:
+                            option.set_value(kvp->get<double>());
+                            break;
+                        case KvpValue::Type::INT64:
+                            option.set_value(kvp->get<int64_t>());
+                            break;
+                        case KvpValue::Type::STRING:
+                            fill_option_from_string_kvp(option, kvp);
+                            break;
+                        case KvpValue::Type::GUID:
+                            fill_option_from_guid_kvp(option, kvp);
+                          break;
+                        default:
+                            return;
+                            break;
+                    }
+                });
+        });
+}
+
+GncOptionDB*
+gnc_get_optiondb_from_dispatcher(SCM dispatcher)
+{
+    SCM  get_options = scm_c_eval_string("gnc:options-get");
+    if (dispatcher == SCM_BOOL_F)
+        return nullptr;
+    auto scm_ptr{scm_call_1(get_options, dispatcher)};
+    auto smob{!scm_is_null(scm_ptr) && SCM_INSTANCEP(scm_ptr) &&
+              scm_is_true(scm_slot_exists_p(scm_ptr, SCM_EOL)) ?
+              scm_slot_ref(scm_ptr, SCM_EOL) : (scm_ptr)};
+
+    void *c_ptr{nullptr};
+    if (!SCM_NULLP(smob))
+    {
+        if (SCM_POINTER_P(smob))
+            c_ptr = SCM_POINTER_VALUE(smob);
+        else
+            c_ptr = reinterpret_cast<void*>(SCM_CELL_WORD_1(smob));
+    }
+    auto u_ptr{static_cast<std::unique_ptr<GncOptionDB>*>(c_ptr)};
+    return u_ptr->get();
+}
```

Comment:
Another example of using SCM code in gnc-optionsdb.

Reply:
There are more below. A simple search on 'SCM' will reveal all.

Reply:
Um, misplaced? There's no SCM there, though `c_ptr` is set from an SCM value. But it's still the same function.

The others are `gnc_register_internal_option` that can certainly be moved into gnc-options.i, and `gnc_option_db_(lookup|set)_scm_value` that are called from dialog-report-column-view that creates the Scheme values from C. That's easily replaced.

Reply:
Not exactly misplaced. It referred to  the whole `gnc_get_optiondb_from_dispatcher` function that uses several SCM specific calls. I didn't reallize it was the same function I already had mentioned in the header file though.